packages/debug/src/browser/debug-configuration-manager.ts

```diff
@@ -153,11 +163,61 @@ export class DebugConfigurationManager {
     get current(): DebugSessionOptions | undefined {
         return this._currentOptions;
     }
+
+    async getSelectedConfiguration(): Promise<DebugSessionOptions | undefined> {
+        // providerType applies to dynamic configurations only
+        if (!this._currentOptions?.providerType) {
+            return this._currentOptions;
+        }
+
+        // Refresh a dynamic configuration from the provider.
+        // This allow providers to update properties before the execution e.g. program
+        const { providerType, configuration: { name } } = this._currentOptions;
+        const configuration = await this.fetchDynamicDebugConfiguration(name, providerType);
+
+        if (!configuration) {
+            const message = nls.localize(
+                'theia/debug/missingConfiguration',
+                "Dynamic configuration '{0}:{1}' is missing or not applicable", providerType, name);
+            throw new Error(message);
+        }
+
+        return { configuration, providerType };
+    }
+
     set current(option: DebugSessionOptions | undefined) {
         this.updateCurrent(option);
+        this.updateRecentlyUsedDynamicConfigurationOptions(option);
     }
+
+    protected updateRecentlyUsedDynamicConfigurationOptions(option: DebugSessionOptions | undefined): void {
+        if (option?.providerType) { // if it's a dynamic configuration option
+            // Removing an item already present in the list
+            const index = this.recentDynamicOptionsTracker.findIndex(item => this.dynamicOptionsMatch(item, option));
+            if (index > -1) {
+                this.recentDynamicOptionsTracker.splice(index, 1);
+            }
+            // Adding new item, most recent at the top of the list
+            const recentMax = 3;
+            if (this.recentDynamicOptionsTracker.unshift(option) > recentMax) {
+                // Keep the latest 3 dynamic configuration options to not clutter the dropdown.
+                this.recentDynamicOptionsTracker.splice(recentMax);
+            }
+        }
+    }
+
+    protected dynamicOptionsMatch(one: DebugSessionOptions, other: DebugSessionOptions): boolean {
+        return one.providerType !== undefined
+        && one.configuration.name === other.configuration.name
+        && one.providerType === other.providerType;
+    }
+
+    get recentDynamicOptions(): DebugSessionOptions[] {
+        return [...this.recentDynamicOptionsTracker];
+    }
```

Comment:
Let's trust our users and just type the return as `readonly DebugSessionOptions[]`:

```suggestion
    get recentDynamicOptions(): readonly DebugSessionOptions[] {
        return this.recentDynamicOptionsTracker;
    }
```